From 1238ba43bcf710573ab04114e52e0bdee2a1497c Mon Sep 17 00:00:00 2001
From: Paul Harrison <pharrison@mythtv.org>
Date: Sat, 8 Feb 2014 19:20:38 +0000
Subject: [PATCH 1/6] MetaIO*: Add the filename to write() and
 writeVolatileMetadata()

This makes them consistent with all the other functions and is also necessary
because the filename stored in the MusicMetadata may no longer point to a file
on the local file system that we can write to.

Note: bumps the ABI core and plugins need to be recompiled.
---
 mythplugins/mythmusic/mythmusic/editmetadata.cpp  |    2 +-
 mythplugins/mythmusic/mythmusic/flacencoder.cpp   |    7 +------
 mythplugins/mythmusic/mythmusic/importmusic.cpp   |    2 +-
 mythplugins/mythmusic/mythmusic/lameencoder.cpp   |    7 +------
 mythplugins/mythmusic/mythmusic/musicplayer.cpp   |    2 +-
 mythplugins/mythmusic/mythmusic/vorbisencoder.cpp |    7 +------
 mythtv/libs/libmythbase/mythversion.h             |    2 +-
 mythtv/libs/libmythmetadata/metaio.cpp            |    2 +-
 mythtv/libs/libmythmetadata/metaio.h              |    8 ++++----
 mythtv/libs/libmythmetadata/metaioavfcomment.cpp  |    2 +-
 mythtv/libs/libmythmetadata/metaioavfcomment.h    |    2 +-
 mythtv/libs/libmythmetadata/metaioflacvorbis.cpp  |    7 +++++--
 mythtv/libs/libmythmetadata/metaioflacvorbis.h    |    2 +-
 mythtv/libs/libmythmetadata/metaioid3.cpp         |   13 +++++++++----
 mythtv/libs/libmythmetadata/metaioid3.h           |    4 ++--
 mythtv/libs/libmythmetadata/metaiomp4.cpp         |    7 ++++---
 mythtv/libs/libmythmetadata/metaiomp4.h           |    2 +-
 mythtv/libs/libmythmetadata/metaiooggvorbis.cpp   |    7 +++++--
 mythtv/libs/libmythmetadata/metaiooggvorbis.h     |    2 +-
 mythtv/libs/libmythmetadata/metaiotaglib.cpp      |    2 +-
 mythtv/libs/libmythmetadata/metaiotaglib.h        |    2 +-
 mythtv/libs/libmythmetadata/metaiowavpack.cpp     |    7 +++++--
 mythtv/libs/libmythmetadata/metaiowavpack.h       |    2 +-
 23 files changed, 50 insertions(+), 50 deletions(-)

diff --git a/mythplugins/mythmusic/mythmusic/editmetadata.cpp b/mythplugins/mythmusic/mythmusic/editmetadata.cpp
index fc02be4..b8a0fd4 100644
--- a/mythplugins/mythmusic/mythmusic/editmetadata.cpp
+++ b/mythplugins/mythmusic/mythmusic/editmetadata.cpp
@@ -206,7 +206,7 @@ void EditMetadataCommon::saveAll()
 
         if (tagger)
         {
-            tagger->write(m_metadata);
+            tagger->write(m_metadata->Filename(), m_metadata);
             delete tagger;
         }
     }
diff --git a/mythplugins/mythmusic/mythmusic/flacencoder.cpp b/mythplugins/mythmusic/mythmusic/flacencoder.cpp
index df9d4a1..e08314b 100644
--- a/mythplugins/mythmusic/mythmusic/flacencoder.cpp
+++ b/mythplugins/mythmusic/mythmusic/flacencoder.cpp
@@ -92,12 +92,7 @@ FlacEncoder::~FlacEncoder()
     }
 
     if (m_metadata)
-    {
-        QString filename = m_metadata->Filename();
-        m_metadata->setFilename(m_outfile);
-        MetaIOFLACVorbis().write(m_metadata);
-        m_metadata->setFilename(filename);
-    }
+        MetaIOFLACVorbis().write(m_outfile, m_metadata);
 }
 
 int FlacEncoder::addSamples(int16_t *bytes, unsigned int length)
diff --git a/mythplugins/mythmusic/mythmusic/importmusic.cpp b/mythplugins/mythmusic/mythmusic/importmusic.cpp
index a778064..474ee5f 100644
--- a/mythplugins/mythmusic/mythmusic/importmusic.cpp
+++ b/mythplugins/mythmusic/mythmusic/importmusic.cpp
@@ -397,7 +397,7 @@ void ImportMusicDialog::addPressed()
             MetaIO *tagger = MetaIO::createTagger(meta->Filename());
             if (tagger)
             {
-                tagger->write(meta);
+                tagger->write(meta->Filename(), meta);
                 delete tagger;
             }
         }
diff --git a/mythplugins/mythmusic/mythmusic/lameencoder.cpp b/mythplugins/mythmusic/mythmusic/lameencoder.cpp
index ea19330..e9390d2 100644
--- a/mythplugins/mythmusic/mythmusic/lameencoder.cpp
+++ b/mythplugins/mythmusic/mythmusic/lameencoder.cpp
@@ -145,12 +145,7 @@ LameEncoder::~LameEncoder()
 
     // Now write the Metadata
     if (m_metadata)
-    {
-        QString filename = m_metadata->Filename();
-        m_metadata->setFilename(m_outfile);
-        MetaIOID3().write(m_metadata);
-        m_metadata->setFilename(filename);
-    }
+        MetaIOID3().write(m_outfile, m_metadata);
 }
 
 int LameEncoder::addSamples(int16_t * bytes, unsigned int length)
diff --git a/mythplugins/mythmusic/mythmusic/musicplayer.cpp b/mythplugins/mythmusic/mythmusic/musicplayer.cpp
index a79b11f..3087213 100644
--- a/mythplugins/mythmusic/mythmusic/musicplayer.cpp
+++ b/mythplugins/mythmusic/mythmusic/musicplayer.cpp
@@ -1269,7 +1269,7 @@ void MusicPlayer::updateVolatileMetadata(void)
 
                 if (tagger)
                 {
-                    tagger->writeVolatileMetadata(getCurrentMetadata());
+                    tagger->writeVolatileMetadata(getCurrentMetadata()->Filename(), getCurrentMetadata());
                     delete tagger;
                 }
             }
diff --git a/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp b/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
index 444f4cc..150d126 100644
--- a/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
+++ b/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
@@ -104,12 +104,7 @@ VorbisEncoder::~VorbisEncoder()
 
     // Now write the Metadata
     if (m_metadata)
-    {
-        QString filename = m_metadata->Filename();
-        m_metadata->setFilename(m_outfile);
-        MetaIOOggVorbis().write(m_metadata);
-        m_metadata->setFilename(filename);
-    }
+        MetaIOOggVorbis().write(m_outfile, m_metadata);
 }
 
 int VorbisEncoder::addSamples(int16_t * bytes, unsigned int length)
diff --git a/mythtv/libs/libmythmetadata/metaio.cpp b/mythtv/libs/libmythmetadata/metaio.cpp
index ff6a1bb..188cdd3 100644
--- a/mythtv/libs/libmythmetadata/metaio.cpp
+++ b/mythtv/libs/libmythmetadata/metaio.cpp
@@ -202,7 +202,7 @@ void MetaIO::readFromFilename(MusicMetadata* metadata)
     QString artist, album, title, genre;
     int tracknum = 0;
 
-    const QString filename = metadata->Filename();
+    const QString filename = metadata->Filename(false);
 
     if (filename.isEmpty())
         return;
diff --git a/mythtv/libs/libmythmetadata/metaio.h b/mythtv/libs/libmythmetadata/metaio.h
index 80bbe88..5b337f5 100644
--- a/mythtv/libs/libmythmetadata/metaio.h
+++ b/mythtv/libs/libmythmetadata/metaio.h
@@ -23,20 +23,20 @@ class META_PUBLIC MetaIO
     /*!
     * \brief Writes all metadata back to a file
     *
+    * \param filename The filename to write metadata to
     * \param mdata A pointer to a MusicMetadata object
     * \returns Boolean to indicate success/failure.
     */
-    virtual bool write(MusicMetadata* mdata) = 0;
+    virtual bool write(const QString &filename, MusicMetadata* mdata) = 0;
 
     /*!
     * \brief Writes rating and playcount back to a file
-    *
+    * \param filename The filename to write metadata to
     * \param mdata A pointer to a MusicMetadata object
     * \returns Boolean to indicate success/failure.
     */
-    virtual bool writeVolatileMetadata(MusicMetadata* mdata)
+    virtual bool writeVolatileMetadata(const QString & /*filename*/, MusicMetadata* /*mdata*/)
     {
-        (void)mdata;
         return false;
     }
 
diff --git a/mythtv/libs/libmythmetadata/metaioavfcomment.cpp b/mythtv/libs/libmythmetadata/metaioavfcomment.cpp
index 8bdddd8..d5b2742 100644
--- a/mythtv/libs/libmythmetadata/metaioavfcomment.cpp
+++ b/mythtv/libs/libmythmetadata/metaioavfcomment.cpp
@@ -26,7 +26,7 @@ MetaIOAVFComment::~MetaIOAVFComment(void)
 /*!
  * \copydoc MetaIO::write()
  */
-bool MetaIOAVFComment::write(MusicMetadata* /*mdata*/)
+bool MetaIOAVFComment::write(const QString & /*filename*/, MusicMetadata* /*mdata*/)
 {
     // Wont implement....
     return false;
diff --git a/mythtv/libs/libmythmetadata/metaioavfcomment.h b/mythtv/libs/libmythmetadata/metaioavfcomment.h
index c817d71..096baaa 100644
--- a/mythtv/libs/libmythmetadata/metaioavfcomment.h
+++ b/mythtv/libs/libmythmetadata/metaioavfcomment.h
@@ -22,7 +22,7 @@ public:
     MetaIOAVFComment(void);
     virtual ~MetaIOAVFComment(void);
 
-    bool write(MusicMetadata* mdata);
+    bool write(const QString &filename, MusicMetadata* mdata);
     MusicMetadata* read(const QString &filename);
 
 private:
diff --git a/mythtv/libs/libmythmetadata/metaioflacvorbis.cpp b/mythtv/libs/libmythmetadata/metaioflacvorbis.cpp
index a074e1f..83d27c6 100644
--- a/mythtv/libs/libmythmetadata/metaioflacvorbis.cpp
+++ b/mythtv/libs/libmythmetadata/metaioflacvorbis.cpp
@@ -47,12 +47,15 @@ TagLib::FLAC::File *MetaIOFLACVorbis::OpenFile(const QString &filename)
 /*!
  * \copydoc MetaIO::write()
  */
-bool MetaIOFLACVorbis::write(MusicMetadata* mdata)
+bool MetaIOFLACVorbis::write(const QString &filename, MusicMetadata* mdata)
 {
     if (!mdata)
         return false;
 
-    m_filename = mdata->Filename(true);
+    if (filename.isEmpty())
+        return false;
+
+    m_filename = filename;
 
     TagLib::FLAC::File *flacfile = OpenFile(m_filename);
 
diff --git a/mythtv/libs/libmythmetadata/metaioflacvorbis.h b/mythtv/libs/libmythmetadata/metaioflacvorbis.h
index 196ce59..c757db0 100644
--- a/mythtv/libs/libmythmetadata/metaioflacvorbis.h
+++ b/mythtv/libs/libmythmetadata/metaioflacvorbis.h
@@ -24,7 +24,7 @@ public:
     MetaIOFLACVorbis(void);
     virtual ~MetaIOFLACVorbis(void);
 
-    bool write(MusicMetadata* mdata);
+    bool write(const QString &filename, MusicMetadata* mdata);
     bool writeAlbumArt(const QString &filename, const AlbumArtImage *albumart);
     bool removeAlbumArt(const QString &filename, const AlbumArtImage *albumart);
 
diff --git a/mythtv/libs/libmythmetadata/metaioid3.cpp b/mythtv/libs/libmythmetadata/metaioid3.cpp
index f048065..077a52a 100644
--- a/mythtv/libs/libmythmetadata/metaioid3.cpp
+++ b/mythtv/libs/libmythmetadata/metaioid3.cpp
@@ -157,9 +157,12 @@ TagLib::ID3v1::Tag* MetaIOID3::GetID3v1Tag(bool create)
 /*!
  * \copydoc MetaIO::write()
  */
-bool MetaIOID3::write(MusicMetadata* mdata)
+bool MetaIOID3::write(const QString &filename, MusicMetadata* mdata)
 {
-    if (!OpenFile(mdata->Filename(), true))
+    if (filename.isEmpty())
+        return false;
+
+    if (!OpenFile(filename, true))
         return false;
 
     TagLib::ID3v2::Tag *tag = GetID3v2Tag();
@@ -844,9 +847,11 @@ bool MetaIOID3::writePlayCount(TagLib::ID3v2::Tag *tag, int playcount)
     return true;
 }
 
-bool MetaIOID3::writeVolatileMetadata(MusicMetadata* mdata)
+bool MetaIOID3::writeVolatileMetadata(const QString &filename, MusicMetadata* mdata)
 {
-    QString filename = mdata->Filename();
+    if (filename.isEmpty())
+        return false;
+
     int rating = mdata->Rating();
     int playcount = mdata->PlayCount();
 
diff --git a/mythtv/libs/libmythmetadata/metaioid3.h b/mythtv/libs/libmythmetadata/metaioid3.h
index 8cde7a0..af2be83 100644
--- a/mythtv/libs/libmythmetadata/metaioid3.h
+++ b/mythtv/libs/libmythmetadata/metaioid3.h
@@ -36,8 +36,8 @@ class META_PUBLIC MetaIOID3 : public MetaIOTagLib
     MetaIOID3(void);
     virtual ~MetaIOID3(void);
 
-    virtual bool write(MusicMetadata* mdata);
-    bool writeVolatileMetadata(MusicMetadata* mdata);
+    virtual bool write(const QString &filename, MusicMetadata* mdata);
+    bool writeVolatileMetadata(const QString &filename, MusicMetadata* mdata);
 
     bool writeAlbumArt(const QString &filename, const AlbumArtImage *albumart);
     bool removeAlbumArt(const QString &filename, const AlbumArtImage *albumart);
diff --git a/mythtv/libs/libmythmetadata/metaiomp4.cpp b/mythtv/libs/libmythmetadata/metaiomp4.cpp
index 744e731..eb6b13e 100644
--- a/mythtv/libs/libmythmetadata/metaiomp4.cpp
+++ b/mythtv/libs/libmythmetadata/metaiomp4.cpp
@@ -27,11 +27,14 @@ MetaIOMP4::~MetaIOMP4(void)
 /*!
  * \copydoc MetaIO::write()
  */
-bool MetaIOMP4::write(MusicMetadata* mdata)
+bool MetaIOMP4::write(const QString &filename, MusicMetadata* mdata)
 {
     if (!mdata)
         return false;
 
+    if (filename.isEmpty())
+        return false;
+
 // Disabled because it doesn't actually work. Better implemented with Taglib
 // when we formally move to 1.6
 
@@ -39,8 +42,6 @@ bool MetaIOMP4::write(MusicMetadata* mdata)
 //     AVFormatParameters* p_params = NULL;
 //     AVInputFormat* p_inputformat = NULL;
 //
-//     QString filename = mdata->Filename();
-//
 //     QByteArray local8bit = filename.toLocal8Bit();
 //     if ((av_open_input_file(&p_context, local8bit.constData(),
 //                             p_inputformat, 0, p_params) < 0))
diff --git a/mythtv/libs/libmythmetadata/metaiomp4.h b/mythtv/libs/libmythmetadata/metaiomp4.h
index ec264ae..82eacbe 100644
--- a/mythtv/libs/libmythmetadata/metaiomp4.h
+++ b/mythtv/libs/libmythmetadata/metaiomp4.h
@@ -19,7 +19,7 @@ class META_PUBLIC MetaIOMP4 : public MetaIO
     MetaIOMP4(void);
     virtual ~MetaIOMP4(void);
 
-    bool write(MusicMetadata* mdata);
+    bool write(const QString &filename, MusicMetadata* mdata);
     MusicMetadata* read(const QString &filename);
 
   private:
diff --git a/mythtv/libs/libmythmetadata/metaiooggvorbis.cpp b/mythtv/libs/libmythmetadata/metaiooggvorbis.cpp
index eefe23c..38ca1f9 100644
--- a/mythtv/libs/libmythmetadata/metaiooggvorbis.cpp
+++ b/mythtv/libs/libmythmetadata/metaiooggvorbis.cpp
@@ -41,12 +41,15 @@ TagLib::Ogg::Vorbis::File *MetaIOOggVorbis::OpenFile(const QString &filename)
 /*!
  * \copydoc MetaIO::write()
  */
-bool MetaIOOggVorbis::write(MusicMetadata* mdata)
+bool MetaIOOggVorbis::write(const QString &filename, MusicMetadata* mdata)
 {
     if (!mdata)
         return false;
 
-    m_filename = mdata->Filename();
+    m_filename = filename;
+
+    if (m_filename.isEmpty())
+        return false;
 
     TagLib::Ogg::Vorbis::File *oggfile = OpenFile(m_filename);
 
diff --git a/mythtv/libs/libmythmetadata/metaiooggvorbis.h b/mythtv/libs/libmythmetadata/metaiooggvorbis.h
index c133c5f..6724b23 100644
--- a/mythtv/libs/libmythmetadata/metaiooggvorbis.h
+++ b/mythtv/libs/libmythmetadata/metaiooggvorbis.h
@@ -24,7 +24,7 @@ class META_PUBLIC MetaIOOggVorbis : public MetaIOTagLib
     MetaIOOggVorbis(void);
     ~MetaIOOggVorbis(void);
 
-    bool write(MusicMetadata* mdata);
+    bool write(const QString &filename, MusicMetadata* mdata);
     MusicMetadata* read(const QString &filename);
 
   private:
diff --git a/mythtv/libs/libmythmetadata/metaiotaglib.cpp b/mythtv/libs/libmythmetadata/metaiotaglib.cpp
index fb55de3..f8e0363 100644
--- a/mythtv/libs/libmythmetadata/metaiotaglib.cpp
+++ b/mythtv/libs/libmythmetadata/metaiotaglib.cpp
@@ -85,7 +85,7 @@ void MetaIOTagLib::ReadGenericMetadata(Tag *tag, MusicMetadata *metadata)
     {
         LOG(VB_GENERAL, LOG_ERR,
             QString("MetaIOTagLib: Failed to read metadata from '%1'")
-                .arg(metadata->Filename()));
+                .arg(metadata->Filename(false)));
     }
 }
 
diff --git a/mythtv/libs/libmythmetadata/metaiotaglib.h b/mythtv/libs/libmythmetadata/metaiotaglib.h
index 1a1f82f..e984a3e 100644
--- a/mythtv/libs/libmythmetadata/metaiotaglib.h
+++ b/mythtv/libs/libmythmetadata/metaiotaglib.h
@@ -23,7 +23,7 @@ class META_PUBLIC MetaIOTagLib : public MetaIO
     MetaIOTagLib(void);
     virtual ~MetaIOTagLib(void);
 
-    virtual bool write(MusicMetadata* mdata) = 0;
+    virtual bool write(const QString &filename, MusicMetadata* mdata) = 0;
     virtual MusicMetadata* read(const QString &filename) = 0;
 
   protected:
diff --git a/mythtv/libs/libmythmetadata/metaiowavpack.cpp b/mythtv/libs/libmythmetadata/metaiowavpack.cpp
index 499f758..435bd09 100644
--- a/mythtv/libs/libmythmetadata/metaiowavpack.cpp
+++ b/mythtv/libs/libmythmetadata/metaiowavpack.cpp
@@ -43,12 +43,15 @@ TagLib::WavPack::File *MetaIOWavPack::OpenFile(const QString &filename)
 /*!
 * \copydoc MetaIO::write()
 */
-bool MetaIOWavPack::write(MusicMetadata* mdata)
+bool MetaIOWavPack::write(const QString &filename, MusicMetadata* mdata)
 {
     if (!mdata)
         return false;
 
-    m_filename = mdata->Filename();
+    if (filename.isEmpty())
+        return false;
+
+    m_filename = filename;
 
     TagLib::WavPack::File *wpfile = OpenFile(m_filename);
 
diff --git a/mythtv/libs/libmythmetadata/metaiowavpack.h b/mythtv/libs/libmythmetadata/metaiowavpack.h
index 3d1e5ab..f024227 100644
--- a/mythtv/libs/libmythmetadata/metaiowavpack.h
+++ b/mythtv/libs/libmythmetadata/metaiowavpack.h
@@ -26,7 +26,7 @@ public:
     MetaIOWavPack(void);
     virtual ~MetaIOWavPack(void);
 
-    bool write(MusicMetadata* mdata);
+    bool write(const QString &filename, MusicMetadata* mdata);
     MusicMetadata* read(const QString &filename);
 
 private:
-- 
1.7.10

